### PR TITLE
xds: implement XdsClient thread-safety and synchronization

### PR DESCRIPTION
**Pull request by voidzcy**
Based on the design of go/separate-client-server-xds-client and #7491, with thread-safe implementation of XdsClient APIs that can be called from different channels. Similar to C-core's implementation https://github.com/grpc/grpc/pull/24200.

Related changes for LoadReportClient: #7528.

### CODE REVIEW

File: xds/src/main/java/io/grpc/xds/ServerXdsClient.java
Comment (sanjaypujare) [inline xds/src/main/java/io/grpc/xds/ServerXdsClient.java:49]: This does not seem to be used anywhere. Does it mean there will be another PR for the server side to actually start using `ServerXdsClient` from the current `XdsClientImpl2`?

Comment (voidzcy) [reply xds/src/main/java/io/grpc/xds/ServerXdsClient.java:49]: Yes, exactly.

File: xds/src/main/java/io/grpc/xds/ServerXdsClient.java
Comment (sanjaypujare) [inline xds/src/main/java/io/grpc/xds/ServerXdsClient.java:82]: Few comments about this sequence of actions:

- if `adjustResourceSubscription` decided to not issue a request because it's in backoffMode do you still need to set the timer?
- why `updateNodeMetadataForListenerRequest` if you are in backoffMode and not going to issue the request?
- can the INFO log statement be issued only if a request is issued (and the timer set)?
- if the ServerXdsClient is going to be shared among multiple server instances then updating the common `node` instance is going to be problematic. Hopefully with the new server side API, modifying Node will not be needed. So we can ignore this problem for now 

I understand some of the above was existing code but I understand one of the goals is to fix whatever issues we can fix.

File: xds/src/main/java/io/grpc/xds/AbstractXdsClient.java
Comment (sanjaypujare) [inline xds/src/main/java/io/grpc/xds/AbstractXdsClient.java:208]: This can be made `abstract` instead of returning `null` as I don't see any advantage of providing a default implementation.

File: xds/src/main/java/io/grpc/xds/ServerXdsClient.java
Comment (sanjaypujare) [inline xds/src/main/java/io/grpc/xds/ServerXdsClient.java:82]: Should it be `if (!isInBackoff())` ?

File: xds/src/main/java/io/grpc/xds/ClientXdsClient.java
Comment (sanjaypujare) [inline xds/src/main/java/io/grpc/xds/ClientXdsClient.java:81]: `ServerXdsClient` will also start using resources so I was thinking this can be moved up into AbstractXdsClient but we can hold off on it.

File: xds/src/main/java/io/grpc/xds/ServerXdsClient.java
Comment (voidzcy) [reply xds/src/main/java/io/grpc/xds/ServerXdsClient.java]: - if `adjustResourceSubscription` decided to not issue a request because it's in backoffMode do you still need to set the timer?

Yeah, this is something missed. It should check before scheduling the timer. It should call `isInBackoff()` to check before scheduling the timer. Good catch. Fixed it.

- why `updateNodeMetadataForListenerRequest` if you are in backoffMode and not going to issue the request?

I think in the existing `XdsClientImpl2` this is wrong. It should indeed update the Node even if currently in backoff and no request is being sent. Currently in backoff just means we should not send the request now, the retry task will send the request once the timer fires. So even if the retry timer is pending, we should set up the Node properly so that once the timer fires it picks up the latest updated Node for sending the request.

- can the INFO log statement be issued only if a request is issued (and the timer set)?

I think it is ok as the log just says it *starts watching*, which means the watcher is added, not necessarily the resource is requested. It's just for business logic level (the caller of this API does not care about how XdsClient subscribe to the resources, it just need to know that its watcher is added and will be notified later). There are other logs for requests being sent.

File: xds/src/main/java/io/grpc/xds/ServerXdsClient.java
Comment (voidzcy) [reply xds/src/main/java/io/grpc/xds/ServerXdsClient.java]: Oops. Sorry, my bad.

File: xds/src/main/java/io/grpc/xds/ClientXdsClient.java
Comment (voidzcy) [reply xds/src/main/java/io/grpc/xds/ClientXdsClient.java:81]: No, resource subscription should belong to the business logic layer. So they should be kept here. Basically, the abstract layer should only abstract away the logic of how it send/receive xDS requests/responses.

File: xds/src/main/java/io/grpc/xds/AbstractXdsClient.java
Comment (voidzcy) [reply xds/src/main/java/io/grpc/xds/AbstractXdsClient.java]: Sure. Made it abstract.

File: xds/src/main/java/io/grpc/xds/ServerXdsClient.java
Comment (sanjaypujare) [reply xds/src/main/java/io/grpc/xds/ServerXdsClient.java]: I highly recommend having a  unit test that checks that `ldsRespTimer` is not created/set if the client is in backoffMode to catch such things.

Comment (voidzcy) [reply xds/src/main/java/io/grpc/xds/ServerXdsClient.java]: That needs to be done for client and server side respectively. On the client side we have [`streamClosedAndRetryRaceWithAddRemoveWatchers`](https://github.com/grpc/grpc-java/blob/0e7cd05bf47d9a408bb834c10558789d231e3175/xds/src/test/java/io/grpc/xds/XdsClientImplTest2.java#L1433). But it doesn't verify the resource subscribe timers. It can be added easily into that test case. But maybe not now. This is an existing behavior, so we should not expect anything related to it to be changed even though it is missing some tests. Also, can you take care of the server side if you want make server side test more robust?

File: xds/src/main/java/io/grpc/xds/AbstractXdsClient.java
Comment (sanjaypujare) [inline xds/src/main/java/io/grpc/xds/AbstractXdsClient.java:252]: Behavior of `getSubscribedResources` as I have understood is:

- the one in `ServerXdsClient` returns null if no watcher set else emptyList
- the one in `ClientXdsClient` returns null instead of emptyList
- over here if it is `null` you change it back to emptyList

Why can't we then have `getSubscribedResources` return emptyList instead of null? So null should never be returned unless I am missing something

Comment (voidzcy) [reply xds/src/main/java/io/grpc/xds/AbstractXdsClient.java:252]: The problem really comes from the `ServerXdsClient`'s implementation. As you mentioned, returning `null` and empty list means different things for `ServerXdsClient`: `null` indicates nothing is being subscribed to and when the ADS stream is recreated, no request should be sent. While empty list does mean it is currently subscribing to the wildcard Listener resource, and a request need to be sent when the ADS stream is recreated.

`ClientXdsClient`'s implementation will always return `null` if nothing being subscribed to, or a non-empty list. It has to be consistent with `ServerXdsClient`'s semantics. It cannot return an empty list as its semantics has alreayd been taken. The abstract layer does not know if the real implementation is a `ServerXdsClient` or a `ClientXdsClient`.

Comment (sanjaypujare) [reply xds/src/main/java/io/grpc/xds/AbstractXdsClient.java:252]: Okay, thanks for the clarification. I thought the client side (ClientXdsClient) also uses wildcard queries (since they are supported for clusters and listeners) but it looks like our client-side never needs to issue wild-card queries because of the specific use-case of proxyless gRPC (unlike Envoy which uses wildcard queries to get all the clusters and listeners in the mesh).

The server side will also now use specific resources (assuming TD implements the new API) so the emptyList case should never happen and this code can then be cleaned up.

File: xds/src/main/java/io/grpc/xds/ServerXdsClient.java
Comment (sanjaypujare) [inline xds/src/main/java/io/grpc/xds/ServerXdsClient.java:83]: Can you just call `handleStreamRestarted()` to eliminate code duplication? because you are using constants etc it will be helpful in eliminating trivial errors.

File: xds/src/main/java/io/grpc/xds/ServerXdsClient.java
Comment (sanjaypujare) [inline xds/src/main/java/io/grpc/xds/ServerXdsClient.java:215]: `ClientXdsClient` has `cleanUpResourceTimers`. We can combine the 2 into 1 abstract method `cleanUpResourceTimers` in `AbstractXdsClient` which might also simplify the code at call sites.

Comment (voidzcy) [reply xds/src/main/java/io/grpc/xds/ServerXdsClient.java:215]: No, `cleanUpResourceTimers` should be concrete implantation's own logic. The abstract layer does not aware how concrete implementations subscribing to resources. This `cleanUpResourceTimer` method is just a helper method in the concrete implementation, it's used for both XdsClient shutdown and `handleStreamClosed` callback.

File: xds/src/main/java/io/grpc/xds/ServerXdsClient.java
Comment (voidzcy) [reply xds/src/main/java/io/grpc/xds/ServerXdsClient.java:83]: No, that's wrong. `handleStreamRestarted()` should be *effectively abstract*, only the abstract layer should call it. Concrete layers should only provides the implementation to the abstract layer.

Also, upon stream recreated, it should only schedule the timer if the resource has not been resolved (as the management server can choose to not send resources already sent in the previous stream). This is a pending change/bugfix.

File: xds/src/test/java/io/grpc/xds/AbstractXdsClientTest.java
Comment (sanjaypujare) [inline xds/src/test/java/io/grpc/xds/AbstractXdsClientTest.java:46]: This has only `MessagePrinter` tests and `MessagePrinter` is only used in log messages and not for any functionality. So these tests don't test any functionality, is that correct?

Comment (voidzcy) [reply xds/src/test/java/io/grpc/xds/AbstractXdsClientTest.java:46]: The XdsClient's behaviors (including functionalities such as ADS stream closed and retry implemented in the abstract laye) are tested in each of the concrete implementations (see `ClientXdsClientTest` and `ServerXdsClientTest`). In high-level, it's more like a blackbox test.

File: xds/src/main/java/io/grpc/xds/ClientXdsClient.java
Comment (dapengzhang0) [inline xds/src/main/java/io/grpc/xds/ClientXdsClient.java:96]: You can delete now, otherwise it will fail to build.

Comment (voidzcy) [reply xds/src/main/java/io/grpc/xds/ClientXdsClient.java]: Sure. Updated.

File: xds/src/main/java/io/grpc/xds/AbstractXdsClient.java
Comment (dapengzhang0) [inline xds/src/main/java/io/grpc/xds/AbstractXdsClient.java:450]: Add an INFO level log for sending discovery `type` request for `resources`? Just as the INFO levels log in NACK and ACK.

File: xds/src/main/java/io/grpc/xds/AbstractXdsClient.java
Comment (dapengzhang0) [inline xds/src/main/java/io/grpc/xds/AbstractXdsClient.java:500]: Provide the type name?

File: xds/src/main/java/io/grpc/xds/AbstractXdsClient.java
Comment (dapengzhang0) [inline xds/src/main/java/io/grpc/xds/AbstractXdsClient.java:522]: The method has almost identical name as `handleRpcStreamClosed()`.
So `propagateStreamError()` might be a better name, as it's propagating error to the watchers.

File: xds/src/main/java/io/grpc/xds/AbstractXdsClient.java
Comment (dapengzhang0) [inline xds/src/main/java/io/grpc/xds/AbstractXdsClient.java:563]: `stubV2` can be moved into `start()` as a local variable.

File: xds/src/main/java/io/grpc/xds/AbstractXdsClient.java
Comment (dapengzhang0) [inline xds/src/main/java/io/grpc/xds/AbstractXdsClient.java:212]: Usages of `adjustResourceSubscription ` seem not synchronized in the sense of `runWithSynchronized()`.

**Review by dapengzhang0 — COMMENTED** [review]
Nice work to abstract out the RPC layer!

File: xds/src/main/java/io/grpc/xds/AbstractXdsClient.java
Comment (voidzcy) [reply xds/src/main/java/io/grpc/xds/AbstractXdsClient.java:212]: Which usage specifically? All of its invocations should have been holding the lock.

But, btw, I am highly suspecting the server side *SynchronizationContext* is wrong. It is already wrong in [XdsClientWrapperForServerSds](https://github.com/grpc/grpc-java/blob/master/xds/src/main/java/io/grpc/xds/XdsClientWrapperForServerSds.java). It doesn't synchronize anything, because only tasks from the XdsClient's channel thread are scheduled into it.

/cc @sanjaypujare You probably do not understand how those threads works.

File: xds/src/main/java/io/grpc/xds/AbstractXdsClient.java
Comment (voidzcy) [reply xds/src/main/java/io/grpc/xds/AbstractXdsClient.java]: Okay, changed, although this constructor is called by `super()` with a new stopwatch gotten from the supplier in concrete classes.

File: xds/src/main/java/io/grpc/xds/AbstractXdsClient.java
Comment (voidzcy) [reply xds/src/main/java/io/grpc/xds/AbstractXdsClient.java]: Thanks.

File: xds/src/main/java/io/grpc/xds/ServerXdsClient.java
Comment (voidzcy) [reply xds/src/main/java/io/grpc/xds/ServerXdsClient.java:93]: Yeah, that's better. Updated.

File: xds/src/main/java/io/grpc/xds/AbstractXdsClient.java
Comment (voidzcy) [reply xds/src/main/java/io/grpc/xds/AbstractXdsClient.java:447]: Sure, added a log message.

File: xds/src/main/java/io/grpc/xds/AbstractXdsClient.java
Comment (voidzcy) [reply xds/src/main/java/io/grpc/xds/AbstractXdsClient.java:498]: Hmm, the enum value doesn't carry the raw type name (aka, typeUrl), this method is already using abstracted information. The real type URL should have been logged in the raw message, so it should be good enough to just warn with the the unknownness.

File: xds/src/main/java/io/grpc/xds/AbstractXdsClient.java
Comment (voidzcy) [reply xds/src/main/java/io/grpc/xds/AbstractXdsClient.java:520]: Hmm, I think `handleStreamClosed()` is better aligned with other "*abstract*" methods. Calling it "propagate" may not be that appropriate as this is actually how inheritance works. It is not directly calling another classes' method,  it is calling itself's method.

File: xds/src/main/java/io/grpc/xds/AbstractXdsClient.java
Comment (voidzcy) [reply xds/src/main/java/io/grpc/xds/AbstractXdsClient.java]: Sure. Cleaned up the `AdsStreamV2` and `AdsStreamV3` naming as well, made the class itself more descriptive.

File: xds/src/main/java/io/grpc/xds/AbstractXdsClient.java
Comment (dapengzhang0) [reply xds/src/main/java/io/grpc/xds/AbstractXdsClient.java:498]: Yeah, we can't print the actual type name.

File: xds/src/main/java/io/grpc/xds/AbstractXdsClient.java
Comment (voidzcy) [reply xds/src/main/java/io/grpc/xds/AbstractXdsClient.java:209]: Basically, calling into XdsClient's API (e.g., `XdsClient.watchListenerData()`) should also be from inside the SynchronizationContext.

File: xds/src/main/java/io/grpc/xds/AbstractXdsClient.java
Comment (dapengzhang0) [reply xds/src/main/java/io/grpc/xds/AbstractXdsClient.java:520]: Not a big deal. Just felt names are too closed.

File: xds/src/main/java/io/grpc/xds/AbstractXdsClient.java
Comment (dapengzhang0) [reply xds/src/main/java/io/grpc/xds/AbstractXdsClient.java:209]: Oh, sorry, I forgot they were inside lock and misunderstood the code.

**Review by dapengzhang0 — APPROVED** [review]
LGTM

**Comment by voidzcy** [comment]
Merging. Has LGTM at least for client side. We can promote `ClientXdsClient` for client side first while keeping XdsClientImpl2 around for server side. If server side needs more changes, we can do it easily in `ServerXdsClient` before promoting it for server side..

**Comment by sanjaypujare** [comment]
> Merging. Has LGTM at least for client side. We can promote `ClientXdsClient` for client side first while keeping XdsClientImpl2 around for server side. If server side needs more changes, we can do it easily in `ServerXdsClient` before promoting it for server side..

I am okay with that. Just to clarify the server side enhancements based on XdsClientImpl2 are ready to be made but will be changed to use ServerXdsClient assuming it's not a big change

**Review by sanjaypujare — APPROVED** [review]
LGTM

**Pull request merged**
